HW #4

ArturAbdullinITIS opened the pull request:
Screencast link: https://disk.360.yandex.ru/i/1y5y_A1B7i-GTQ

ArturAbdullinITIS committed:
Delete .idea directory

ArturAbdullinITIS committed:
feat: implement 4th homework

-Add MainScreen with viewModel
-Add UseCase to work with coroutines
-Add mapper to simplify work with ui
-Add hilt to work with DI

ArturAbdullinITIS committed:
fix: text using string resources

ArturAbdullinITIS committed:
Merge remote-tracking branch 'origin/feature/homework4' into feature/homework4

ArturAbdullinITIS committed:
fix: wrong naming for cancelled coroutines count toast

GrKnght commented on app/src/main/java/ru/itis/practice/domain/CoroutinesInteractor.kt:
```diff
@@ -0,0 +1,105 @@
+import java.io.IOException
+import javax.inject.Inject
+
+class CoroutinesUseCase @Inject constructor(
```
То, что я сейчас распишу - штука дискуссионная, в разных командах бывает разный нейминг. Расскажу про то, как принято у нас.
Класс называется `UseCase`-ом, если он делает единичную операцию. Т.е. launchAllCoroutines - первое действие, cancelAllCoroutines - второе, getCompletedCount - третий. Если вы хотите оставаться в этом контексте, тогда следует поделить все на 3 юз кейса.
Если же класс выполняет какую-то совокупность действий в одном контексте, тогда следует называть его Interactor-ом. В вашем случае CoroutinesInteractor, который соберет все действия по работе с корутинами

ArturAbdullinITIS replied:
Да, я думал над этим, тоже не хотел называть use case'ом, так как не 1 операцию выполняет, но не придумал альтернативы. Теперь понял как в таком случае стоит называть, спасибо!

GrKnght commented on app/src/main/java/ru/itis/practice/domain/CoroutinesUseCase.kt:
```diff
@@ -0,0 +1,105 @@
+        onProgress: (completed: Int, total: Int) -> Unit,
+        onException: (Exception) -> Unit,
+    ): List<String> =
+        coroutineScope {
```
Не очень понятно, почему вы запускаете отдельный скоуп. Если разворачивать код, у вас получается примерно так:
```
viewModelScope.launch{
   coroutineScope {
      ...
   }
}
```
Есть случаи, да, когда нам нужен новый скоуп внутри скоупа. Но это больше исключения. Более распространенный кейс, когда нужно вернуть какой-то результат - это использование `withContext`-а
```
viewModelScope.launch {
   val result = withContext(selectedDispatchers) {
      ...
   }
}
```

GrKnght commented on app/src/main/java/ru/itis/practice/domain/CoroutinesInteractor.kt:
```diff
@@ -0,0 +1,105 @@
+    }
+
+
+    private fun throwRandException(index: Int): Exception {
```
По условию задания, вы должны выбрасывать exception. Сейчас вы просто создаете экземпляр экзепшена и возвращаете его
Под выбрасыванием имеется в виду вызов оператора
```
throw ClassNotFoundException(...) 
```
Сейчас у вас нет обработки ошибок, поэтому выбрасывание будет ронять приложение, нужно код немного доработать.
И условие номер 2 - Exception должен выбрасываться с 30% шансом. 
Т.е. сначала 30 процентов на то, что выполнение корутины займет 7 и более секунд, и еще 30 процентов на то, что выкинется какой-то экзепшен. А у вас при выполнении более 7 секунд всегда возвращается ошибка

ArturAbdullinITIS replied:
Видимо неправильно понял задание, исправлю.

GrKnght commented on app/src/main/java/ru/itis/practice/domain/CoroutinesUseCase.kt:
```diff
@@ -0,0 +1,105 @@
+import javax.inject.Inject
+
+class CoroutinesUseCase @Inject constructor(
+    @ApplicationContext private val context: Context
```
Мы про это еще не говорили, но если прямо с точки зрения архитектуры подходить, то в классы с бизнес логикой, по-хорошему, контекст напрямую лучше не прокидывать. Бывают случаи, когда без этого не обойтись. Но сейчас контекст используется для получения ресурсов.
Лучше создать новый класс, тот же ResManager, который будет возвращать вам нужные ресурсы. А внутри по капотом этот ResManager уже и будет использовать appContext для получения ресурсов

GrKnght reviewed (commented):
Работа проверена, выполнено крепко.

UPD:
Посмотрел правки, все хорошо

ArturAbdullinITIS committed:
fix: exception handling problems,
minor misnaming, scope issues

-Add ResManager class to provide string resources instead of providing application context directly into domain layer
-Update coroutine launch method to handle exceptions and not creating new extra scope
